### ISSUE TITLE
Large throughput performance regression

### ISSUE THREAD

**Issue by timothysc** [GoogleCloudPlatform/kubernetes#8119]
**Test:**
Running a very simple wide guestbook (1000 pods) on a 23 node bare metal cluster, with a side script continually calling 'kubectl get pods' once a second to parse the state. 

Note: This is similar to many other tests we've ran in the past, and the model we would like to add to the e2e data dumping. 

---
**Environment:** 
Atomic 7.1.1 bare metal cluster with 23-nodes. 
GitVersion:"v0.16.2-649-g738f403eea96d4", 

---
**Expected Results:**
Linear damped impulse response curve with consistent slope (throughput).  We previously had seen consistent numbers. 

Before regression 4/22/2015:
![image](https://cloud.githubusercontent.com/assets/169553/7593162/9ca84724-f89d-11e4-81ae-7373ce064b69.png)

---
**Actual Results: 5/12/2015**

![image](https://cloud.githubusercontent.com/assets/169553/7592559/9a35bfa2-f899-11e4-9fc9-a804befbce46.png)

histogram - bad distribution: 
![image](https://cloud.githubusercontent.com/assets/169553/7592822/59519900-f89b-11e4-86ec-b92df3751bbd.png)

--- 
**Other Notes:**
Anecdotal: 
- It now appears that garbage collection is batched, when it does run it seems to grind the apiserver to a halt as seen in the graphs.  Previous version would show spikes but they were not as high and did not affect overall performance.  

My initial suspicions are the apiserver changes, but I need more proof.  

/cc @wojtek-t @davidopp @fgrzadkowski @smarterclayton @rrati @jayunit100

refs: #7573, #7790

**Comment by bprashanth** [comment]
> consistent numbers ~15 pods/sec

So previously you had 1000 pods go from 0->running in a minute? Did you bump up the rate limit restrictions on your rest clients? 

Currently it will take us roughly that long to even create 1000 pods through the default client, so that red line capping at ~60s is expected IIUC. The kubelets use a 5qps connection, so to start 40 pods/node you're probably looking at another 1/2 minute in pure request time client side.

These aside, a slow list is the main cause of our density tests failing today (https://github.com/GoogleCloudPlatform/kubernetes/issues/7561).

**Comment by timothysc** [comment]
> consistent numbers ~15 pods/sec

The added b4 graph shows the actual results before regression.

**Comment by davidopp** [comment]
@lavalamp

**Comment by lavalamp** [comment]
... But if I do math on your chart, the total scheduling took 4'30" in the "regression" case but 5'30" in the "good" case... how exactly is this a regression? why do you care about the slope?

**Comment by lavalamp** [comment]
Wait, nvm... I'm doing math on the wrong parts of the chart.

**Comment by lavalamp** [comment]
So it went from 4'15" to 4'30", if I read the labels correctly this time.

**Comment by lavalamp** [comment]
@timothysc Can you label the axes on the "bad distribution" chart? I have no idea what that is.

**Comment by lavalamp** [comment]
This is a 6% slowdown... does that really warrant the "large" moniker and P0 tag?

**Comment by bprashanth** [comment]
Hmm, the time to get a 1000 running pods roughly went from 2m -> 4m right? In the first one it took 45s to create all the pods, in the second, it took 1.20. This is expected (rc ratelimiting, also explains the tiny blip @11.30 in the second one), and we can speed it up at the cost of random things acting funny in the cluster.
 
so the only unexplained blip to me is at 11:20:57 in the second graph, when the rc creates the remaining 500 pods, why does the running pods graph not grow linearly like it did for the first 500? Note that this is also 43 non pause pods per node, and we're tracking memory/cpu pressure on kubelet in a few other bugs (https://github.com/GoogleCloudPlatform/kubernetes/issues/7815).

**Comment by davidopp** [comment]
I interpreted it as going from ~2.5m to ~4m to get the tasks scheduled (i.e. out of pending) while end-to-end the amount of time to get them running remained around ~4m.

**Comment by bprashanth** [comment]
It would also help to know which ones are pending+assigned vs just pending

**Comment by lavalamp** [comment]
Yeah you guys are right, I cannot interpret charts today.

**Comment by lavalamp** [comment]
@timothysc 
> It now appears that garbage collection is batched

Garbage collection? Like from the go runtime? I don't think that changed?

> My initial suspicions are the apiserver changes, but I need more proof.

What apiserver changes?

My first suspicion was the changes to rc manager; however, as @bprashanth points out, that seems to have been a net improvement, now creating pods only slightly faster than the system can start them, at least for the first minute or so.

**Comment by lavalamp** [comment]
@timothysc since it's a cheap experiment, can you set [BurstReplicas](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/controller/replication_controller.go#L69) to something >= 1000 and try again? This will rule out or implicate recent changes to RC manager.

**Comment by rrati** [comment]
In the pre-regression case, pods start running at about 15:54:58 and all are running at about 15:57:04, which is a total start time of about 2 minutes.

In the regression case, pods start running at about 11:19:22 and all are running at about 11:23:27, which is a total start time of about 4 minutes.

**Comment by lavalamp** [comment]
Yeah I will go get lunch and ponder how I can possibly be so bad at reading charts now.

**Comment by davidopp** [comment]
ping @timothysc -- did you have a chance to do what @lavalamp suggested here?
https://github.com/GoogleCloudPlatform/kubernetes/issues/8119#issuecomment-101389859

**Comment by bprashanth** [comment]
Fyi depending on the value you increase it by, you will also want to tweak the expiration on the cache  per the comment above it describing expected qps/latencies 
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/controller/replication_controller.go#L61

It sounds like all you're doing is measuring the time it takes to push 1000 pods, 40 per node, into running.   I plan to run some experiments along those lines today.

**Comment by lavalamp** [comment]
If it turns out scheduler is a bottleneck, I have at least one planned
optimization that I haven't been prioritizing.

**Comment by fgrzadkowski** [comment]
What do you mean by garbage collection?

-- sent from a mobile device
12 maj 2015 18:39 "Timothy St. Clair" <notifications@github.com> napisał(a):

**Comment by timothysc** [comment]
@fgrzadkowski  

\#perf top -p \`pgrep kube-apiserver\`

I had observed that the plateaus in the graph correlated with large spikes in: 
kube-apiserver  [.] scanblock
kube-apiserver  [.] runtime.mallocgc
kube-apiserver  [.] runtime.MSpan_Sweep

**Comment by fgrzadkowski** [comment]
My guess is that it's the time when we have spikes of list pod requests coming from kubelets. You'd observe high CPU usage that affects request latency and overall performance.

For the past days I was investigating possible things to decrease CPU usage. Currently I'm looking into using a different JSON parse:
https://github.com/ugorji/go/tree/master/codec
I hope to use it in etcd client library, our internal parsing and DeepCopy. Initial benchmarks are very promising (e.g. DeepCopy ~6x faster).

Btw - what master do you use for your experiments? How many cores?

**Comment by timothysc** [comment]
@fgrzadkowski our nodes are bare metal 40-core blades. Intel(R) Xeon(R) CPU E5-2680 v2 @ 2.80GHz

**Comment by timothysc** [comment]
@lavalamp re: BurstReplicas - no benefit.

**Comment by bprashanth** [comment]
@timothysc how I would love your logs between 11:20->23
do you have the BurstReplicas=1000 graph?

**Comment by lavalamp** [comment]
OK, that confirms that controller manager is probably not the cause.

**Comment by timothysc** [comment]
@bprashanth  BurstReplicas=1000 

![image](https://cloud.githubusercontent.com/assets/169553/7620571/72a4fe38-f986-11e4-8962-be2c93502764.png)

**Comment by bprashanth** [comment]
hello flatline.

It takes roughly 2m to push 1000 pause pods across 100 nodes to running. This correlates well with our 1.0 goal to be linear, since it takes 5-6m to push 3000 pausers acrosss the same number of nodes (we know this from our jenkins runs). 

So at face value, the biggest difference seems to be that you're using non-pause images, and moving more of them per node than kube is used to, though I suspect a combination of fun things is really happening.

**Comment by lavalamp** [comment]
OK so the problem is clearly not RC manager. So the question is, is it apiserver, scheduler, or what? what changes after 600 pods are running? Random hypothesis: is it the storm from all the kubelets updating statuses at the same time overloads the scheduler's watch? But I don't think that has changed.

@timothysc can you post logs of apiserver & scheduler so we can maybe see which one is the culprit?

(PS @bprashanth it doesn't matter if it's linear-- I care about throughput and monotonicity, but not linearity.)

**Comment by wojtek-t** [comment]
>
If it turns out scheduler is a bottleneck, I have at least one planned
optimization that I haven't been prioritizing.

@lavalamp - I think it's not 100% true. Basically pods are scheduling quite quickly, but the problem is that sending a binding to apiserver is done synchronously and is blocking scheduling the next pod. Which means that scheduler is spending most of its time waiting for the response from apiserver.

**Comment by timothysc** [comment]
The question I have is: what **fundamental behavior**, at the system level, has changed from 4/22 to now.

**Comment by bprashanth** [comment]
> (PS @bprashanth it doesn't matter if it's linear-- I care about throughput and monotonicity, but not linearity.)

IIRC our 1.0 goals were 99 percentil pod startup time should be < X for a given pod, node configuration (30 pause pods on 100 nodes) and it should scale linearly after that.

**Comment by lavalamp** [comment]
@wojtek-t

>@lavalamp - I think it's not 100% true. Basically pods are scheduling quite quickly, but the problem is that sending a binding to apiserver is done synchronously and is blocking scheduling the next pod. Which means that scheduler is spending most of its time waiting for the response from apiserver.

That's the thing that my planned optimization would fix.

**Comment by lavalamp** [comment]
@wojtek-t It's faster to do it than to explain it: #8261

**Comment by lavalamp** [comment]
@timothysc if you're up for more experiments, you could try running with #8261.

**Comment by wojtek-t** [comment]
>
@wojtek-t It's faster to do it than to explain it: #8261

:) thanks - will take a look today.

**Comment by fgrzadkowski** [comment]
@timothysc 

I synced today to HEAD and did some profiling and for the whole density test we spend 30% of our time in ```crypto/tls.(*serverHAndshakeState).doFullHandshake```. Has something changed recently regarding authorization? I know we've been trying to get rid of kubernetes-ro. Maybe that's related? @lavalamp You've been working on this. Any ideas?

**Comment by smarterclayton** [comment]
This seems like we're not reusing connections.  Did we change to not reuse the underlying TCP transport pool?  Did someone change something in how we connect to etcd?

----- Original Message -----

**Comment by timothysc** [comment]
@fgrzadkowski I don't have tls on for our test cluster, been waiting on #8261 to land.  I'll just pull and rebuild.

**Comment by lavalamp** [comment]
@fgrzadkowski I did change the network test to do TLS and will be changing more in the future. But I don't think that can explain 30%. @smarterclayton's hypothesis seems like a good starting point (maybe kubelet's client isn't reusing?).

**Comment by timothysc** [comment]
@lavalamp w/ #8261 it's slightly better overall.   

![image](https://cloud.githubusercontent.com/assets/169553/7659628/013c812e-fb09-11e4-9c4a-ae3b2e546ae5.png)

but still a off quite a bit from 4/22.

**Comment by bprashanth** [comment]
Well 4/22 would be better too with the scheduler fix, because goroutines

**Comment by lavalamp** [comment]
So that shaved a minute off, which is nice.

Thoughts:

My money is on some interaction between apiserver & kubelets.

"running" or not is computed by kubelet and done via write-- it's possible there's been no system change other than a delay or change there.

It's possible the etcd cache is suffering lock contention? (unlikely)

Listing all pods once per second is (unfortunately) an amount of load on the system that's larger than all the kubelets in your cluster put together. If the test were retried with a watch instead of hammering the server with polls, the result might look very different.

How many cores does etcd have in these tests?

Would be super interesting to know what changes at ~620 pods in. What is causing that threshold? If we turn off (say) 5 nodes in the cluster, does that inflection point go down, stay the same, or increase?

**Comment by wojtek-t** [comment]
>
Listing all pods once per second is (unfortunately) an amount of load on the system that's larger than all the kubelets in your cluster put together. If the test were retried with a watch instead of hammering the server with polls, the result might look very different.

I completely agree - many people took a lot of effort to eliminate number of lists all over the code. We should also change the test to eliminate it.

>
How many cores does etcd have in these tests?

I don't know the answer for this question, but we were playing with in in the context of our Density tests so I can give a quick update on it.
Basically, currently we are using 1 core for etcd (with 4-core master machine). When we tried to use more, it was significantly slowing down the whole system (calls to etcd had lower latency, but apiserver was significantly slower and we had a hypothesis that golang scheduler wasn't doing a particularly good job when scheduling the threads). Giving more cores to etcd made the situation even worse.

However, I think that this is related to apiserver - it is currently too loaded. Once we improve it, we will be able to give etcd more cores and this will improve the latency of requests to etcd.

**Comment by timothysc** [comment]
@lavalamp 

>Listing all pods once per second is (unfortunately) an amount of load on the system that's larger 
>than all the kubelets in your cluster put together. If the test were retried with a watch instead of 
>hammering the server with polls, the result might look very different.
The list has little effect when we stagger the intervals.  It appears to be interrupted by other events see details below.  

>How many cores does etcd have in these tests?
40

When we graph and store our runs in prometheus, it shows the following interesting indicators: 
1. The endpoint list appears to have much greater impact.  apiserver_request_latencies_summary{verb="LIST"}

2. From etcd_request_latencies_summary we see endpoints, namespace, node, and events seem to dominate api.pod is lower on the list.   

If you have dashboard recipes I'd be happy to share, b/c that seems like a better route to diagnose the root cause.

**Comment by lavalamp** [comment]
@timothysc endpoints? That's interesting. Can you publish the entire list? maybe time to look at kube-proxy...

**Comment by timothysc** [comment]
Sorry for my lag on this ticket...

@lavalamp every plateau will correspond 1:1 with large lags in the endpoint controller: 

 Finished syncing service "default/redis-master" endpoints. (1m30.116068618s)...  

I've not dug into the details why yet.

**Comment by lavalamp** [comment]
@timothysc when you say etcd has 40 cores, that means you set its GOMAXPROCS env var to 40, right?

Also, I suspect the causality is slow apiserver -> slow endpoint sync, not the other way around.

**Comment by timothysc** [comment]
@lavalamp correct.

**Comment by wojtek-t** [comment]
>
Also, I suspect the causality is slow apiserver -> slow endpoint sync, not the other way around.

I agree this is probably true.

@timothysc - did you run those experiments recently synced to head? I think a bunch of thing were improved recently so the question is whether it's still a problem.

**Comment by timothysc** [comment]
@wojtek-t yes, we(myself, @rrati, & @jayunit100) resynch to the tip of master nightly.

**Comment by bgrant0607** [comment]
Do we have any idea how much read vs. write etcd traffic we generate? An option for scaling the read load would be etcd proxies: https://coreos.com/etcd/docs/2.0.8/proxy.html

**Comment by bprashanth** [comment]
I looked at this at some point. IIRC both reads and writes are still proxied to the core cluster but the proxy will shuffle around the member it hits. With the default consistency settings this could lead to weirdness.

**Comment by timothysc** [comment]
After running multiple experiments today, the plateaus still exist, but it looks much more like a storm as outlined in #4521 from the nodes stampeding, but many lists are affected by a storm of writes from the nodes. 

  
![image](https://cloud.githubusercontent.com/assets/169553/7891315/55d73720-0611-11e5-82d8-45fa870f9499.png)

**Comment by lavalamp** [comment]
@timothysc can you label the axes, please? What unit does "M" represent?

**Comment by lavalamp** [comment]
Also what are the different colored lines?

**Comment by bprashanth** [comment]
> If you graph etcd latencies, they all increase from the write spike from the nodes.

Can you turn off events and try? I think there's a flag to do it but I just comment out the recorder line. 
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/client/record/event.go#L258

**Comment by timothysc** [comment]
@lavalamp the graph on the left is a prom-dash of: apiserver_request_latencies_summary{verb="LIST"} 
The graph on the right is a prom-dash of: 
apiserver_request_latencies_summary{verb=~"PUT|POST"}
I also plot: 
etcd_request_latencies_summary

same scale for the vars from /metrics.

@bprashanth The graph on the right is *put|posts the spike is only from .99 quantile resource:node verb:put.

The node put spike precedes all other lags.

**Comment by bgrant0607** [comment]
Any idea what caused the spike in node writes? I think the apiserver log should show what requests it received.

**Comment by timothysc** [comment]
@bprashanth well I seem to have spoke a little to soon, it cuts the spike in the lists in about 1/2.

**Comment by bprashanth** [comment]
@timothysc did it affect the running pods response curve? Maybe it's time to profile a cluster where we send events to a different etcd (this was considered previously, but at the time we had bigger bottlenecks). The default out of box cluster would still use a single store.

There are improvements to the etcd api underway which will make batched pod creates easier to achieve post 1.0.

**Comment by timothysc** [comment]
@bprashanth The slopes appear better but it still has long plateaus. 

![image](https://cloud.githubusercontent.com/assets/169553/7893601/06b8ce60-0625-11e5-9050-fc41e0900224.png)

**Comment by lavalamp** [comment]
@timothysc I'm sorry but I still can't make heads or tails of your chart. On the right, something spiked to 60M, what unit is that in? And on the left, is each colored line a different request? or if not, what do they represent? And it goes up to 50M, what does that mean? Presumably the X axis is time?

**Comment by jayunit100** [comment]
the 60M is the latency (60 milliseconds) of requests during that time slice.   So what its saying is "at this time period, the latency was 60 seconds".

**Comment by lavalamp** [comment]
In that case, those numbers really don't make much sense to me, are you sure? 60ms average sounds better than I'd expect even with no load-- I thought our average api call took about 100ms. Maybe it's because you're running this on way beefier systems than I usually do. How are exactly are you measuring these?

What about requests that took longer than a second?

And finally what are the different colored lines? :) :)

(BTW, median, 90th, and 99th %iles are better to track than average. Average can get fooled by a few really bad ones.)

**Comment by jayunit100** [comment]
Oh ! And the colored lines, of course :) ![screenshot from 2015-05-29 20 03 25](https://cloud.githubusercontent.com/assets/826111/7894737/de59e044-063d-11e5-8ecd-e8adee2d9c46.png) heres how they work.  Tim's filtering on **LIST** operations.  SO for example, each colored line represents a particular type of list call.  in this case i highlighted one such call (list replicationControllers)

**Comment by lavalamp** [comment]
@timothysc I made #9031 as a wild guess about your problem; could you give it a try when you feel like giving this another timeslice?

**Comment by davidopp** [comment]
@timothysc since we're trying to close out the 1.0 issues, it would be great if you can prioritize helping us with this.

**Comment by timothysc** [comment]
@davidopp ack, I'm looking @ #9031 now.

**Comment by timothysc** [comment]
So I dug through all the logs and it appears that the kcm & kubelets have watch errors in their logs which seem to proceed the plateau: 

 Failed to watch *api.Node: Get http://host18-rack11:8080/api/v1beta3/watch/nodes?fieldSelector=metadata.name%3Dhost14-rack10.scale.openstack.engineering.redhat.com&resourceVersion=203923776: dial tcp 10.1.4.82:8080: connection refused
Jun 02 15:37:20 host14-rack10.scale.openstack.engineering.redhat.com kubelet[18612]: I0602 15:37:20.357038   18612 reflector.go:241] Watch close - *api.Pod total 0 items received
Jun 02 15:37:20 host14-rack10.scale.openstack.engineering.redhat.com kubelet[18612]: I0602 15:37:20.356975   18612 iowatcher.go:102] Unexpected EOF during watch stream event decoding: unexpected EOF
Jun 02 15:37:20 host14-rack10.scale.openstack.engineering.redhat.com kubelet[18612]: I0602 15:37:20.356842   18612 reflector.go:241] Watch close - *api.Service total 0 items received
Jun 02 15:37:20 host14-rack10.scale.openstack.engineering.redhat.com kubelet[18612]: I0602 15:37:20.356778   18612 iowatcher.go:102] Unexpected EOF during watch stream event decoding: unexpected EOF
Jun 02 15:37:20 host14-rack10.scale.openstack.engineering.redhat.com kubelet[18612]: I0602 15:37:20.355713   18612 reflector.go:241] Watch close - *api.Node total 10 items received
Jun 02 15:37:20 host14-rack10.scale.openstack.engineering.redhat.com kubelet[18612]: I0602 15:37:20.355590   18612 iowatcher.go:102] Unexpected EOF during watch stream event decoding: unexpected EOF

^ This appears all over the cluster.

**Comment by bprashanth** [comment]
It sounds like a thundering herd relist problem https://github.com/GoogleCloudPlatform/kubernetes/issues/4521#issuecomment-106114691, those log lines show the watch on your services, pods and node expiring. But from the described latency graph your lists are taking < 1s (https://github.com/GoogleCloudPlatform/kubernetes/issues/8119#issuecomment-106960908)? in which case it is puzzling that this would have any effect on the cluster. 

I'd like to see how many concurrent lists you have, so when you have time can you apply https://github.com/GoogleCloudPlatform/kubernetes/pull/8260 and rerun your test? you will need to modify your test a little to invoke the request monitor, eg: https://github.com/GoogleCloudPlatform/kubernetes/pull/8260/files#diff-e98c4edf3baa4df8ca5b39bd5e7577ddR121. You can set the "LIST" count to something like 30, just to see how many are in flight.

You can also try setting https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/apiserver/apiserver.go#L138 to something high. My in flight pr should do this for production clusters (https://github.com/GoogleCloudPlatform/kubernetes/pull/8964), e2e clusters will still have it low.

Can you also upload your master componenets logs (at verbosity level > 3) somewhere?

**Comment by timothysc** [comment]
@bprashanth Jay was off my a factor of ten, it was >60 seconds from what I was seeing.  

> Finished syncing service "default/redis-master" endpoints. (1m30.116068618s)...

I'll rerun the tests in the a.m. and post the logs.

**Comment by bprashanth** [comment]
That makes a little more sense, also fyi the numbers in `/metrics` are in microseconds :)
Can you give me the exact test you're running, is it just the guestbook example (the one that uses redis), or do you have a script I can play around with?

**Comment by davidopp** [comment]
@timothysc ping? Can you pretty please give the info @bprashanth requested?

**Comment by timothysc** [comment]
@bprashanth the test is simply a very wide slave with a script that pools the pod-status once a second, which has no effect when we change the polling interval.   I'm recording the results and will post soon.

**Comment by bprashanth** [comment]
Ok, it sounds like you just modified the guestbook example to have a 1000 slaves and ran it on a cluster with 30 nodes. I will try and repro it sometime today/tomorrow.

**Comment by timothysc** [comment]
@davidopp @bprashanth @lavalamp log results - https://drive.google.com/file/d/0Bzq0eFfVveNqTWNOQUJrcHlBMGM/view?usp=sharing

**Comment by lavalamp** [comment]
Thanks. Taking a look...

**Comment by lavalamp** [comment]
I processed the apiserver log and did some digging.

It looks like some mystery event happened twice during the run; twice I see clusters of extremely high latency requests. One cluster ends at 14:58:23.513357, the other at 15:00:55.505686.

So between 14:58:13 and 14:58:23, a ton of high latency requests all end. The timing makes me suspicious that they were all blocked on the same thing. There's no real pattern of verb or path, which makes me suspect that etcd is the common denominator. The controller manager log from that time period seems to be doing mostly nothing, and the scheduler log is cut off (or maybe the scheduler is totally done by then).

So, @timothysc, can you also post the etcd log from this run? Hopefully you still have it as it's a bit of a pain to process the logs like I did.

Other findings: whatever manages persistent volumes & claims is list/watching with a 10 second period. We should find it and increase-- it probably doesn't need to be so short.

**Comment by davidopp** [comment]
> Other findings: whatever manages persistent volumes & claims is list/watching with a 10 second period. 

Should we open a PR saying that all the controllers should be taking advantage of #8964 ? I assume the reason you're seeing 10 second period is because using the longer connection timeout is "voluntary" ?

(and just now @bprashanth also pointed me to #8700)

**Comment by lavalamp** [comment]
> Should we open a PR saying that all the controllers should be taking advantage of #8964 ? I assume the reason you're seeing 10 second period is because using the longer connection timeout is "voluntary" ?

An issue, yes-- and yes.

**Comment by bprashanth** [comment]
I'm less afraid of relists from something that doesn't scale with number of nodes. @timothysc how did you setup your cluster? if you used our e2e setup scripts you aren't going to get the longer connection timeout David mentioned.

**Comment by lavalamp** [comment]
Since I just analyzed the log file, I'm pretty confident that setting a longer relist period won't have an effect on @timothysc's test. Only three nodes had watches end during the test.

**Comment by timothysc** [comment]
@bprashanth we provisioned our cluster, and use defaults for most settings.

**Comment by lavalamp** [comment]
So by my math--which was admittedly pretty bad earlier in this thread--you're getting a throughput of about 3 pods per second. I don't think we actually have a 1.0 requirement for throughput, arguably we should, but I think 3pps would meet the hypothetical requirement if we had it. So I think we can kick this out of the 1.0 list. (Even though I think we should be closer to 10pps throughput.)

I continue to think, based on the log file, that the problem is something along the lines of etcd holding up a bunch of requests periodically. (Flushing to disk?)

Ideas to make further progess on this:
* A script from you guys that repros the issue on a GCE cluster.
* Access to your cluster.
* You guys could do a `git bisect` to narrow down when this broke.

**Comment by timothysc** [comment]
@lavalamp - re: etcd, that was my suspecion the entire time.  I'm planning on making some configuration changes on those machines just to see what happens.. similar to details outlined here: https://youtu.be/12ZLq2pHWsI

**Comment by timothysc** [comment]
Update: 
We've recently updated our cluster to an HA configuration and are load-balancing our apiservers.  In this environment we see no plateaus, just a clean curve.  

![image](https://cloud.githubusercontent.com/assets/169553/8218113/8f8979f2-1506-11e5-82d5-3c90ba45a1b5.png)

which leads me to believe there is a artificial limitation in the apiserver, not etcd.

**Comment by lavalamp** [comment]
Do you have etcd in an HA configuration, too? I'll be quite surprised if the relevant difference is actually only multiple apiservers. If you set the number of apiservers in your configuration to 1, does the problem come back?

**Comment by timothysc** [comment]
@lavalamp etcd was always in HA configuration, even before.  load-balancing apiservers was the only real change.

**Comment by jayunit100** [comment]
FYI, these plots now automated in #11003

**Comment by timothysc** [comment]
@lavalamp @wojtek-t - I'm going to close this one out given that 1.0 is out, and try to really around next performance targets on throughput.  

If we intend to support basic batch, then overall throughput is going to matter a lot. 
/cc @smarterclayton
